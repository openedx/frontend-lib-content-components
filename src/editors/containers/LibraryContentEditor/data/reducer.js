import { createSlice } from '@reduxjs/toolkit';
import { modes } from '../constants';
import { StrictDict } from '../../../utils';

const initialState = {
  libraries: {},
  selectedLibraryId: null,
  savedLibraryId: null,
  settings: {
    // [libraryId]: initialLibrarySettings,
    //  This reducer structure allows selected settings
    //  to persist when user switches between libraries.
  },
};

const initialLibrarySettings = {
  version: null,
  mode: modes.random.value,
  count: -1,
  showReset: false,
  blocks: [],               // [ { id: 'id', display_name: 'name', block_type: 'type' } ]
  candidates: [],           // ['id1', 'id2']
  v1BlockRequests: {},      // { [blockId]: 'RequestkeyState' }
};

const library = createSlice({
  name: 'library',
  initialState,
  reducers: {
    initializeFromBlockValue: (state, { payload }) => ({
      ...state,
      savedLibraryId: payload.libraryId,
      selectedLibraryId: payload.libraryId,
      settings: payload.settings,
    }),
    addLibraries: (state, { payload }) => ({
      ...state,
      libraries: {
        ...state.libraries,
        ...payload.libraries,
      },
    }),
    initialLibrarySettings: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [payload.selectedLibraryId]: initialLibrarySettings,
      },
    }),
    setLibraryId: (state, { payload }) => ({
      ...state,
      selectedLibraryId: payload.selectedLibraryId,
    }),
    setLibraryVersion: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          version: payload.version,
        },
      },
    }),
    setModeForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          mode: payload.mode,
        },
      },
    }),
    setCountForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          count: payload.count,
        },
      },
    }),
    setShowResetForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          showReset: payload.showReset,
        },
      },
    }),
    setLibraryBlocks: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          blocks: payload.blocks,
        },
      },
    }),
    setCandidatesForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          candidates: payload.candidates,
        },
      },
    }),
    setV1BlockRequests: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          v1BlockRequests: payload.v1BlockRequests,
        },
      },
    }),
    updateV1BlockRequestStatus: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          v1BlockRequests: {
            ...state.settings[state.selectedLibraryId].v1BlockRequests,
            [payload.blockId]: payload.status,
          },
        },
      },
    }),
    addLibraryBlock: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          blocks: [
            ...state.settings[state.selectedLibraryId].blocks,
            payload.block,
          ],
        },
      },
    }),
  },
});

const actions = StrictDict(library.actions);

const { reducer } = library;

export {
  actions,
  initialState,
  initialLibrarySettings,
  reducer,
};
