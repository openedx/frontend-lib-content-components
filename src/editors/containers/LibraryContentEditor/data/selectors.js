import { createSelector } from 'reselect';
import { modes } from '../constants';
import { getUsageKey } from '../utils';
import * as module from './selectors';

export const libraryState = (state) => state.library;

const mkSimpleSelector = (cb) => createSelector([module.libraryState], cb);

export const simpleSelectors = {
  libraries: mkSimpleSelector(library => library.libraries),
  savedLibraryId: mkSimpleSelector(library => library.savedLibraryId),
  selectedLibraryId: mkSimpleSelector(library => library.selectedLibraryId),
  settings: mkSimpleSelector(library => library.settings),
};

export const blocks = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.blocks ?? [];
    }
    return [];
  },
);

export const v1BlockRequests = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.v1BlockRequests ?? {};
    }
    return [];
  },
);

export const candidates = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.candidates ?? [];
    }
    return [];
  },
);

export const mode = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.mode ?? modes.random.value;
    }
    return modes.random.value;
  },
);

export const libraryPayload = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    let manual = false;
    let shuffle = true;
    let count = null;
    if (selectedLibraryId && settings[selectedLibraryId]) {
      count = settings[selectedLibraryId].count ?? 1;
      if (settings[selectedLibraryId].mode === modes.selected.value) {
        manual = true;
        shuffle = false;
        count = -1;
      }
    }
    return {
      libraryId: selectedLibraryId,
      libraryVersion: settings[selectedLibraryId]?.version,
      manual,
      shuffle,
      count,
      showReset: settings[selectedLibraryId]?.showReset,
      candidates: settings[selectedLibraryId]?.candidates || [],
    };
  },
);

export default {
  ...simpleSelectors,
  blocks,
  candidates,
  mode,
  libraryPayload,
  v1BlockRequests,
};
