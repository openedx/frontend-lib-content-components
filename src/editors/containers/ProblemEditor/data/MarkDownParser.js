/*
This is the direct motivation code taken from:
https://github.com/open-craft/capa-visual-editor/

Changes have been made to adapt the parser with respect to our use case.
*/

import { ProblemTypeKeys } from "../../../data/constants/problem";

export const indexToLetterMap = [...Array(26)].map((val, i) => String.fromCharCode(i + 65));

function getHints(markdown) {
  if (!markdown || !markdown.trim()) {
    return [];
  }
  const hints = [];

  function getHint(row) {
    let hint = '';
    if (row.startsWith('||')) {
      const hintStart = row.indexOf('||');
      const hintEnd = row.slice(hintStart + 2).indexOf('||');
      hint = row.slice(hintStart + 2, hintEnd + 2);
    }
    return hint;
  }
  for (const d in markdown.split('\n')) {
    const row = markdown.split('\n')[d];
    const hint = getHint(row);
    if (hint.length) {
      hints.push({
        id: Number(d),
        value: hint,
      });
    }
  }
  return hints;
}

function getShortAnswerOptions(markdown) {
  // TODO: Include tolerence for number it fails for now.
  let problemType = ProblemTypeKeys.TEXTINPUT;
    if ((!markdown || !markdown.trim())) {
      return {
        shortAnswersList: [],
        problemType,
      };
    }
    const markdownListData = markdown.split('\n');
    const shortAnswersList = [];
    let gotAnswer = false;
    for (const i in markdownListData) {
      const row = markdownListData[i].trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      let feedback = '';
      let answer = '';
      let correct = true;
      if (row.startsWith('=') && !gotAnswer) {
        gotAnswer = true;
        if (feedbackStart > -1) {
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
          answer = row.slice(1, feedbackStart).trim();
        } else {
          answer = row.slice(1).trim();
        }
      } else if (row.startsWith('=') && gotAnswer) {
        break;
      } else if (gotAnswer && row.startsWith('or=')) { // the next correct answer
        if (feedbackStart > -1) {
          answer = row.slice(3, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(3).trim();
        }
      } else if (gotAnswer && row.startsWith('not=')) { // the incorrect answer
        if (feedbackStart > -1) {
          answer = row.slice(4, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(4).trim();
        }
        correct = false;
      }
      if (answer) {
        problemType = isFinite(answer) ? ProblemTypeKeys.NUMERIC : ProblemTypeKeys.TEXTINPUT;
        shortAnswersList.push({
          id: indexToLetterMap[shortAnswersList.length],
          title: answer,
          feedback,
          correct,
        });
      }
    }
    return {
      shortAnswersList,
      problemType,
    };
}

function getMultipleChoiceOptions(markdown) {
  const multipleChoiceOptions = [];
  const data = {
    multiSelectAnswersList: multipleChoiceOptions,
    problemType: ProblemTypeKeys.MULTISELECT,
    groupFeedbackList: [],
  };
  if (!markdown || !markdown.trim()) {
    return data;
  }
  const markdownListData = markdown.split('\n');
  let feedback = '';
  for (const d in markdownListData) {
    const row = markdownListData[d];
    // multiple choices
    if (row.startsWith('[ ]') || row.startsWith('[x]')) {
      // title
      let title = row.slice(row.indexOf(']') + 1).trim();

      const feedbacksStart = row.indexOf('{{');
      const selectedFeedbackStart = row.indexOf(' selected:') || row.indexOf('{selected:'); // two ways to define selected feedback
      const selectedFeedbackEnd = row.slice(selectedFeedbackStart + 1).indexOf('}') + selectedFeedbackStart;
      const unselectedFeedbackStart = row.indexOf('unselected:');
      const unselectedFeedbackEnd = row.slice(unselectedFeedbackStart + 1).indexOf('}}') + unselectedFeedbackStart;

      let selectedFeedback = '';
      let unselectedFeedback = '';
      let titleChanged = false;
      if (feedbacksStart !== -1 && selectedFeedbackStart !== -1) {
        selectedFeedback = row.slice(selectedFeedbackStart + 10, selectedFeedbackEnd);
        title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        titleChanged = true;
      }
      if (feedbacksStart !== -1 && unselectedFeedbackStart !== -1) {
        unselectedFeedback = row.slice(unselectedFeedbackStart + 12, unselectedFeedbackEnd + 1);
        if (!titleChanged) {
          title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        }
      }
      multipleChoiceOptions.push({
        id: indexToLetterMap[multipleChoiceOptions.length],
        title,
        correct: !row.startsWith('[ ]'),
        selectedFeedback,
        unselectedFeedback,
      });
    }
    // group Feedback
    if (multipleChoiceOptions.length) {
      if (row.startsWith('{{')) {
        const groupFeedbackEnd = row.indexOf('}}');
        const goupStart = row.indexOf('((');
        const groupEnd = row.indexOf('))');
        const group = row.slice(goupStart + 2, groupEnd).split(' ');
        const groupChoices = [];
        for (const i in group) {
          const letter = group[i];
          if (multipleChoiceOptions[indexToLetterMap.indexOf(letter)]) {
            groupChoices.push(multipleChoiceOptions[indexToLetterMap.indexOf(letter)].id);
          }
        }
        feedback = row.slice(groupEnd + 2, groupFeedbackEnd).trim();

        if (groupChoices.length && feedback) {
          data.groupFeedbackList.push({
            id: groupChoices.length,
            answers: groupChoices,
            feedback,
          });
        }
      }
    }
  }
  return data;
}

function getSingleChoiceOptions(markdown) {
  // TODO: Build support for selected and unselected feedback
  if (!markdown || !markdown.trim()) {
    return {
      singleSelectAnswersList: [],
      problemType: ProblemTypeKeys.DROPDOWN,
    };
  }

  const singleChoiceOptions = [];
  let dropDownMode = false;
  const markdownListData = markdown.split('\n');
  for (const d in markdownListData) {
    const row = markdownListData[d].trim();
    if (row.startsWith('[[')) {
      dropDownMode = true;
      continue;
    }
    if (row.startsWith(']]')) {
      break;
    }
    if (dropDownMode) {
      if (row.startsWith('(')) {
        let feedback = '';
        let title = row.slice(row.indexOf('(') + 1, row.indexOf(')')).trim();
        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = title.slice(0, feedbackStart).trim();
        }
        const correct = true;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      } else {
        let feedback = '';
        let title = '';
        // let title = row.slice(row.indexOf(')') + 1).trim();

        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = row.slice(0, feedbackStart).trim();
        } else {
          title = row.trim();
        }
        const correct = false;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      }
    } else if (row.startsWith('( )') || row.startsWith('(x)')) {
      let feedback = '';
      let title = row.slice(row.indexOf(')') + 1).trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
        feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
        title = title.slice(0, feedbackStart - 4).trim();
      }
      const correct = !!row.startsWith('(x)');
      singleChoiceOptions.push({
        id: indexToLetterMap[singleChoiceOptions.length],
        title,
        correct,
        feedback,
      });
    }
  }

  return {
    singleSelectAnswersList: singleChoiceOptions,
    problemType: dropDownMode ? ProblemTypeKeys.DROPDOWN : ProblemTypeKeys.SINGLESELECT,
  };
}

function getEditorData(markdown) {
  // TODO: Prase heading in this function.
  if (!markdown || !markdown.trim()) {
    return '';
  }
  let description = '';
  const markdownData = markdown.trim().split('\n');
  for (const i in markdownData) {
    const row = markdownData[i];
    if (['{', '(', '[', '=', '|'].indexOf(row.trim()[0]) === -1) {
      const questionTextStart = row.indexOf('>>');
      const questionTextEnd = row.indexOf('<<');
      let questionText = row;
      if (questionTextStart !== -1 && questionTextEnd !== -1 && (questionTextEnd > questionTextStart)) {
        questionText = row.replace('>>', '<p>').replace('<<', '</p>');
      }
      description += `${questionText}\n`;
    } else {
      break;
    }
  }
  return description;
}

function parseMarkdown(markdown) {
  if (!markdown || !markdown.trim()) {
    return {};
  }
  let answers = {};
  let problemType = '';
  const question = getEditorData(markdown);
  const hints = getHints(markdown);
  let data = {question, settings: {hints}};
  let parsedData = getSingleChoiceOptions(markdown);
  if (parsedData.singleSelectAnswersList.length !== 0){
    answers = parsedData.singleSelectAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  parsedData = getMultipleChoiceOptions(markdown);
  if (parsedData.multiSelectAnswersList.length !== 0){
    answers = parsedData.multiSelectAnswersList;
    problemType = parsedData.problemType;
    const groupFeedbackList = parsedData.groupFeedbackList;
    data = {...data, answers, problemType, groupFeedbackList};
  }
  parsedData = getShortAnswerOptions(markdown);
  if (parsedData.shortAnswersList.length !== 0){
    answers = parsedData.shortAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  return data;
}

function getScorringSettings() {
  return {
    selectedAttemptsOption: window.LXCData.max_attempts,
    selectedPointOption: window.LXCData.weight,
  };
}

export {
  getMultipleChoiceOptions,
  getSingleChoiceOptions,
  getEditorData,
  getHints,
  getScorringSettings,
  getShortAnswerOptions,
  parseMarkdown,
};
