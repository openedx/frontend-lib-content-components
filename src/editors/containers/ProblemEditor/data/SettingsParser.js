import _ from 'lodash-es';

import { ShowAnswerTypes } from '../../../data/constants/problem';

export const popuplateItem = (parentObject, itemName, statekey, metadata) => {
  let parent = parentObject;
  const item = _.get(metadata, itemName, null);
  if (!_.isNil(item)) {
    parent = { ...parentObject, [statekey]: item };
  }
  return parent;
};

export const parseScoringSettings = (metadata) => {
  let scoring = {};

  let attempts = popuplateItem({}, 'max_attempts', 'number', metadata);
  if (!_.isEmpty(attempts)) {
    const unlimited = _.isNaN(attempts.number);
    attempts = { ...attempts, unlimited };
    scoring = { ...scoring, attempts };
  }

  scoring = popuplateItem(scoring, 'weight', 'weight', metadata);

  return scoring;
};

export const parseShowAnswer = (metadata) => {
  let showAnswer = {};

  const showAnswerType = _.get(metadata, 'showanswer', {});
  if (!_.isNil(showAnswerType) && showAnswerType in ShowAnswerTypes) {
    showAnswer = { ...showAnswer, on: showAnswerType };
  }

  showAnswer = popuplateItem(showAnswer, 'attempts_before_showanswer_button', 'afterAttempts', metadata);

  return showAnswer;
};

export const parseSettings = (metadata) => {
  let settings = {};

  if (_.isNil(metadata) || _.isEmpty(metadata)) {
    return settings;
  }

  settings = popuplateItem(settings, 'matlab_api_key', 'matLabApiKey', metadata);

  const scoring = parseScoringSettings(metadata);
  if (!_.isEmpty(scoring)) {
    settings = { ...settings, scoring };
  }

  const showAnswer = parseShowAnswer(metadata);
  if (!_.isEmpty(showAnswer)) {
    settings = { ...settings, showAnswer };
  }
  settings = popuplateItem(settings, 'show_reset_button', 'showResetButton', metadata);
  settings = popuplateItem(settings, 'submission_wait_seconds', 'timeBetween', metadata);

  return settings;
};
