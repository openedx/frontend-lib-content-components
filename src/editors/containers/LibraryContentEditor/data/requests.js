import { StrictDict } from '../../../utils';
import { RequestKeys } from '../../../data/constants/requests';
import { networkRequest } from '../../../data/redux/thunkActions/requests';

/* eslint-disable import/no-cycle */
import { selectors } from '../../../data/redux';
import api from './api';
import apiMethods from '../../../data/services/cms/api';

/**
 * Tracked fetchV1Libraries api method.
 * Tracked to the `fetchV1Libraries` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV1Libraries = ({ ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV1Libraries,
    promise: api.fetchV1Libraries({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV2Libraries api method.
 * Tracked to the `fetchV2Libraries` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV2Libraries = ({ ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV2Libraries,
    promise: api.fetchV2Libraries({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV2LibraryContent api method.
 * Tracked to the `fetchV2LibraryContent` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV1LibraryContent = ({ libraryId, ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV1LibraryContent,
    promise: api.fetchV1LibraryContent({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      libraryId,
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV2LibraryContent api method.
 * Tracked to the `fetchV2LibraryContent` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV2LibraryContent = ({ libraryId, ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV2LibraryContent,
    promise: api.fetchV2LibraryContent({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      libraryId,
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV1LibraryBlock api method.
 * Tracked to the `fetchV1LibraryBlock` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV1LibraryBlock = ({ blockId, ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV1LibraryBlock,
    promise: apiMethods.fetchBlockById({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      blockId,
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchChildrenInfo api method.
 * Tracked to the `fetchChildrenInfo` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchChildrenInfo = ({ ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchChildrenInfo,
    promise: api.fetchChildrenInfo({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      blockId: selectors.app.blockId(getState()),
    }),
    ...rest,
  }));
};

export default StrictDict({
  fetchV1Libraries,
  fetchV2Libraries,
  fetchV1LibraryContent,
  fetchV2LibraryContent,
  fetchV1LibraryBlock,
  fetchChildrenInfo,
});
