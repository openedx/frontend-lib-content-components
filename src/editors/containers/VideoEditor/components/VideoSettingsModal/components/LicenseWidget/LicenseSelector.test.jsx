import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import { formatMessage } from '../../../../../../../testUtils';
import { actions, selectors } from '../../../../../../data/redux';
import { LicenseSelector, mapStateToProps, mapDispatchToProps } from './LicenseSelector';

jest.mock('react', () => {
  const updateState = jest.fn();
  return {
    ...jest.requireActual('react'),
    updateState,
    useContext: jest.fn(() => ({ license: ['error.license', jest.fn().mockName('error.setLicense')] })),
  };
});

jest.mock('react-redux', () => {
  const dispatchFn = jest.fn();
  return {
    ...jest.requireActual('react-redux'),
    dispatch: dispatchFn,
    useDispatch: jest.fn(() => dispatchFn),
  };
});

jest.mock('../../../../../../data/redux', () => ({
  actions: {
    video: {
      updateField: jest.fn().mockName('actions.video.updateField'),
    },
  },
  selectors: {
    video: {
      courseLicenseType: jest.fn(state => ({ courseLicenseType: state })),
    },
  },
}));

describe('LicenseSelector', () => {
  const props = {
    intl: { formatMessage },
    license: 'all-rights-reserved',
    level: 'course',
    courseLicenseType: 'all-rights-reserved',
    updateField: jest.fn().mockName('args.updateField'),
  };
  describe('snapshots', () => {
    test('snapshots: renders as expected with default props', () => {
      expect(
        shallow(<LicenseSelector {...props} />).snapshot,
      ).toMatchSnapshot();
    });
    test('snapshots: renders as expected with library level', () => {
      expect(
        shallow(<LicenseSelector {...props} level="library" />).snapshot,
      ).toMatchSnapshot();
    });
    test('snapshots: renders as expected with block level', () => {
      expect(
        shallow(<LicenseSelector {...props} level="block" />).snapshot,
      ).toMatchSnapshot();
    });
    test('snapshots: renders as expected with no license', () => {
      expect(
        shallow(<LicenseSelector {...props} license="" />).snapshot,
      ).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('courseLicenseType from video.courseLicenseType', () => {
      expect(
        mapStateToProps(testState).courseLicenseType,
      ).toEqual(selectors.video.courseLicenseType(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    test('updateField from actions.video.updateField', () => {
      expect(mapDispatchToProps.updateField).toEqual(dispatch(actions.video.updateField));
    });
  });
});
