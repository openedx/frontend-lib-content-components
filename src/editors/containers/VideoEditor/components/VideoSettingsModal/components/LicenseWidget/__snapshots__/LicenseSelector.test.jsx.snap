// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicenseSelector snapshots snapshots: renders as expected with block level 1`] = `
<Fragment>
  <ActionRow>
    <Form.Control
      as="select"
      className="w-100 m-0 p-0"
      defaultValue="all-rights-reserved"
      disabled={false}
      floatingLabel="License Type"
      onChange={[Function]}
    >
      <option
        hidden={true}
      >
        Select
      </option>
      <option
        value="all-rights-reserved"
      >
        All Rights Reserved
      </option>
      <option
        value="creative-commons"
      >
        Creative Commons
      </option>
    </Form.Control>
    <Fragment>
      <ActionRow.Spacer />
      <IconButtonWithTooltip
        iconAs="Icon"
        onClick={[Function]}
        tooltipContent={
          <FormattedMessage
            defaultMessage="Clear and apply the course-level license"
            description="Message presented to user for action to delete license selection"
            id="authoring.videoeditor.license.deleteLicenseSelection"
          />
        }
        tooltipPlacement="top"
      />
    </Fragment>
  </ActionRow>
  <div
    className="x-small mt-3"
  >
    <FormattedMessage
      defaultMessage="This license is set specifically for this video"
      description="Helper text for license type when choosing for a spcific video"
      id="authoring.videoeditor.license.defaultLevelDescription.helperText"
    />
  </div>
  <div
    className="border-primary-100 mt-3 border-bottom"
  />
</Fragment>
`;

exports[`LicenseSelector snapshots snapshots: renders as expected with default props 1`] = `
<Fragment>
  <ActionRow>
    <Form.Control
      as="select"
      className="w-100 m-0 p-0"
      defaultValue="all-rights-reserved"
      disabled={true}
      floatingLabel="License Type"
      onChange={[Function]}
    >
      <option
        hidden={true}
      >
        Select
      </option>
      <option
        value="all-rights-reserved"
      >
        All Rights Reserved
      </option>
      <option
        value="creative-commons"
      >
        Creative Commons
      </option>
    </Form.Control>
  </ActionRow>
  <div
    className="x-small mt-3"
  >
    <FormattedMessage
      defaultMessage="This license currently set at the course level"
      description="Helper text for license type when using course license"
      id="authoring.videoeditor.license.courseLevelDescription.helperText"
    />
  </div>
  <div
    className="border-primary-100 mt-3 border-bottom"
  />
</Fragment>
`;

exports[`LicenseSelector snapshots snapshots: renders as expected with library level 1`] = `
<Fragment>
  <ActionRow>
    <Form.Control
      as="select"
      className="w-100 m-0 p-0"
      defaultValue="all-rights-reserved"
      disabled={false}
      floatingLabel="License Type"
      onChange={[Function]}
    >
      <option
        hidden={true}
      >
        Select
      </option>
      <option
        value="all-rights-reserved"
      >
        All Rights Reserved
      </option>
      <option
        value="creative-commons"
      >
        Creative Commons
      </option>
    </Form.Control>
    <Fragment>
      <ActionRow.Spacer />
      <IconButtonWithTooltip
        iconAs="Icon"
        onClick={[Function]}
        tooltipContent={
          <FormattedMessage
            defaultMessage="Clear and apply the course-level license"
            description="Message presented to user for action to delete license selection"
            id="authoring.videoeditor.license.deleteLicenseSelection"
          />
        }
        tooltipPlacement="top"
      />
    </Fragment>
  </ActionRow>
  <div
    className="x-small mt-3"
  >
    <FormattedMessage
      defaultMessage="This license currently set at the library level"
      description="Helper text for license type when using library license"
      id="authoring.videoeditor.license.libraryLevelDescription.helperText"
    />
  </div>
  <div
    className="border-primary-100 mt-3 border-bottom"
  />
</Fragment>
`;

exports[`LicenseSelector snapshots snapshots: renders as expected with no license 1`] = `
<Fragment>
  <ActionRow>
    <Form.Control
      as="select"
      className="w-100 m-0 p-0"
      defaultValue=""
      disabled={true}
      floatingLabel="License Type"
      onChange={[Function]}
    >
      <option
        hidden={true}
      >
        Select
      </option>
      <option
        value="all-rights-reserved"
      >
        All Rights Reserved
      </option>
      <option
        value="creative-commons"
      >
        Creative Commons
      </option>
    </Form.Control>
  </ActionRow>
  <div
    className="x-small mt-3"
  >
    <FormattedMessage
      defaultMessage="This license currently set at the course level"
      description="Helper text for license type when using course license"
      id="authoring.videoeditor.license.courseLevelDescription.helperText"
    />
  </div>
  <div
    className="border-primary-100 mt-3 border-bottom"
  />
</Fragment>
`;
